forem/forem
# Introduce domain model for campaigns

_I tagged entirely too many people in this PR, please don't feel like you have to comment. My intention was just to make people aware of the idea and to maybe spark a general discussion about ensuring that important domain concepts don't fall through the cracks but get proper representation._

## What type of PR is this? (check all applicable)


## Description

Campaigns are clearly a concept in our app's domain, as we quite frequently talk about them. However, so far this domain concept had no actual representation in code but instead existed "between the lines" in `SiteConfig` and code in the view layer. This PR introduces a plain Ruby object (PORO) to remedy that. This not only allows us to treat campaigns as first-class citizens in code but also to use niceties like decorators, etc.

**Notes**

1. This was a 1h spike, so I did not change all instances of `SiteConfig.campaign_*` yet. Instead, I changed just a little bit to demonstrate the approach and will change the rest if we agree to move forward with this. The same for tests.
2. I personally always treated `app/models` as the home for all domain models, whether they are DB-backed or not (which IMHO is an implementation detail). Some people seem to be uncomfortable with having non-AR models there. I can move it if people feel better about it, but welcome suggestions regarding the new location.
3. I won't insist on the singleton pattern, but from what I can tell we can only have one active campaign at a time, so it made sense to me.

## Related Tickets & Documents

n/a

## QA Instructions, Screenshots, Recordings

Configure a campaign via `/internal/config`:

<img width="1118" alt="Screen Shot 2020-07-14 at 10 01 35" src="https://user-images.githubusercontent.com/47985/87378860-0911bb00-c5b9-11ea-88b6-6e2815da6b12.png">

Make sure to tick the checkbox next to `Campaign sidebar enabled` and have a look at the sidebar, both with and without a campaign URL set.

## Added tests?

- [x] not yet because spike

## Added to documentation?
